lib/ipam/ipam_block_reader_writer.go

```diff
@@ -160,25 +178,51 @@ func isPoolInRequestedPools(pool cnet.IPNet, requestedPools []cnet.IPNet) bool {
 	return false
 }
 
-func (rw blockReaderWriter) claimBlockAffinity(ctx context.Context, subnet cnet.IPNet, host string, config IPAMConfig) error {
-	// Claim the block affinity for this host.  See model.BlockAffinityValue
-	// for details on the hard-coded value that is used.
-	log.Infof("Host %s claiming block affinity for %s", host, subnet)
+// getPendingAffinity claims a pending affinity for the given host and subnet. The affinity can then
+// be used to claim a block. If an affinity already exists, it will attempt to mark the existing affinity
+// as pending.
+func (rw blockReaderWriter) getPendingAffinity(ctx context.Context, host string, subnet cnet.IPNet) (*model.KVPair, error) {
+	// Make sure hostname is not empty.
+	if host == "" {
+		log.Errorf("Hostname can't be empty")
+		return nil, errors.New("Hostname must be sepcified to claim block affinity")
+	}
+
+	log.Infof("Host %s claiming pending affinity for %s", host, subnet)
 	obj := model.KVPair{
 		Key:   model.BlockAffinityKey{Host: host, CIDR: subnet},
-		Value: model.BlockAffinityValue,
+		Value: model.BlockAffinity{State: model.StatePending},
 	}
-	_, err := rw.client.Create(ctx, &obj)
-
-	// Create the new block.
-	block := newBlock(subnet)
+	aff, err := rw.client.Create(ctx, &obj)
+	if err != nil {
+		if _, ok := err.(cerrors.ErrorResourceAlreadyExists); !ok {
+			log.WithError(err).Error("Failed to claim affinity")
+			return nil, err
+		}
+		log.Info("Block affinity already exists")
 
-	// Make sure hostname is not empty.
-	if host == "" {
-		log.Errorf("Hostname can't be empty")
-		return errors.New("Hostname must be sepcified to claim block affinity")
+		// Get the existing affinity and write it as pending.
+		aff, err = rw.client.Get(ctx, obj.Key, "")
+		if err != nil {
+			log.WithError(err).Error("Failed to get existing affinity")
+			return nil, err
+		}
+		aff.Value.(*model.BlockAffinity).State = model.StatePending
+		return rw.client.Update(ctx, aff)
 	}
+	return aff, nil
+}
+
+// claimBlockAffinity claims the provided block using the given pending affinity. If successful, it will confirm the affinity. If another host
+// steals the block, claimBlockAffinity will attempt to delete the provided pending affinity.
+func (rw blockReaderWriter) claimBlockAffinity(ctx context.Context, aff *model.KVPair, config IPAMConfig) (*model.KVPair, error) {
```

Comment:
I've pulled out the claiming of the affinity block into `getPendingAffinity` - calling functions now need to make sure they have the affinity in-hand in order to call this.

I've also made this return the block that it creates (or the existing one if it exists). So that the caller can use it to CAS an allocation.